test/opt/inline_test.cpp

```diff
@@ -0,0 +1,1071 @@
+     "%color2 = OpVariable %_ptr_Function_v4float Function",
+      "%param = OpVariable %_ptr_Function_v4float Function",
+     "%color3 = OpVariable %_ptr_Function_v4float Function",
+         "%39 = OpLoad %25 %t2D",
+         "%40 = OpLoad %27 %samp",
+         "%41 = OpSampledImage %29 %39 %40",
+         "%42 = OpImageSampleImplicitLod %v4float %41 %32",
+               "OpStore %color1 %42",
+         "%43 = OpLoad %v4float %BaseColor",
+               "OpStore %param %43",
+         "%68 = OpAccessChain %_ptr_Function_float %param %uint_0",
+         "%69 = OpLoad %float %68",
+               "OpStore %66 %69",
+         "%70 = OpLoad %float %66",
+         "%71 = OpFOrdLessThan %bool %70 %float_0",
+               "OpSelectionMerge %72 None",
+               "OpBranchConditional %71 %73 %72",
+         "%73 = OpLabel",
+         "%74 = OpLoad %float %66",
+         "%75 = OpFNegate %float %74",
+               "OpStore %66 %75",
+               "OpBranch %72",
+         "%72 = OpLabel",
+         "%76 = OpLoad %float %66",
+               "OpStore %67 %76",
+         "%44 = OpLoad %float %67",
+         "%45 = OpCompositeConstruct %v4float %44 %44 %44 %44",
+               "OpStore %color2 %45",
+         "%46 = OpLoad %25 %t2D",
+         "%47 = OpLoad %27 %samp",
+         "%77 = OpSampledImage %29 %39 %40",
+         "%48 = OpImageSampleImplicitLod %v4float %77 %35",
+               "OpStore %color3 %48",
+         "%49 = OpLoad %v4float %color1",
+         "%50 = OpLoad %v4float %color2",
+         "%51 = OpFAdd %v4float %49 %50",
+         "%52 = OpLoad %v4float %color3",
+         "%53 = OpFAdd %v4float %51 %52",
+         "%54 = OpCompositeConstruct %v4float %float_3 %float_3 %float_3 %float_3",
+         "%55 = OpFDiv %v4float %53 %54",
+               "OpStore %FragColor %55",
+               "OpReturn",
+               "OpFunctionEnd",
+   "%foo_vf4_ = OpFunction %float None %19",
+        "%bar = OpFunctionParameter %_ptr_Function_v4float",
+         "%56 = OpLabel",
+          "%r = OpVariable %_ptr_Function_float Function",
+         "%57 = OpAccessChain %_ptr_Function_float %bar %uint_0",
+         "%58 = OpLoad %float %57",
+               "OpStore %r %58",
+         "%59 = OpLoad %float %r",
+         "%60 = OpFOrdLessThan %bool %59 %float_0",
+               "OpSelectionMerge %61 None",
+               "OpBranchConditional %60 %62 %61",
+         "%62 = OpLabel",
+         "%63 = OpLoad %float %r",
+         "%64 = OpFNegate %float %63",
+               "OpStore %r %64",
+               "OpBranch %61",
+         "%61 = OpLabel",
+         "%65 = OpLoad %float %r",
+               "OpReturnValue %65",
+               "OpFunctionEnd",
+      // clang-format on
+  };
+  SinglePassRunAndCheck<opt::InlinePass>(JoinAllInsts(before),
+                                                 JoinAllInsts(after),
+                                                 /* skip_nop = */ false);
+}
+
```

Comment:
I'd be ok adding TODOs for this in the test file, and adding the actual tests in subsequent CLs.

SPIR-V core allows recursion I believe all Khronos client APIs ban recursion.  I'd say that handling the recursive case can be deferred: file an enhancement issue for it.